### ISSUE TITLE
Does Form support file upload?

### ISSUE THREAD

**Issue by dodyg** [OrchardCMS/OrchardCore#6027]
I didn't see anything in the Workflow nor the Form widget support for "multipart/form-data"

**Comment by dodyg** [comment]
Or more importantly does the workflow support file upload?

**Comment by deanmarcussen** [comment]
There is no support for file upload through workflows right now.

Related issues https://github.com/OrchardCMS/OrchardCore/issues/4896

**Comment by dodyg** [comment]
OK - let me try to implement a task for the workflow that handle a file upload.

**Comment by dodyg** [comment]
I am trying to write a simple Activity first. This one will write a message to the disk when it is done.

![disk-writer](https://user-images.githubusercontent.com/1761249/80284669-26ed2380-8720-11ea-98ec-8d813dd7e1a2.png)

**Comment by dodyg** [comment]
This is a good simple activity task to learn from https://github.com/EtchUK/Etch.OrchardCore.Workflows/tree/master/FormOutput

**Comment by dodyg** [comment]
If anyone need similar functionality, this code will work

```
  public class DiskWriterTask : TaskActivity
    {
        readonly IOptions<ShellOptions> _shellOptions;
        readonly IStringLocalizer S;
        readonly ShellSettings _shellSettings;
        private readonly IHttpContextAccessor _http;

        public DiskWriterTask(IStringLocalizer<DiskWriterTask> s, 
            IOptions<ShellOptions> shellOptions, 
            ShellSettings shellSettings,
            IHttpContextAccessor httpContextAccessor)
        {
            _shellOptions = shellOptions;
            _shellSettings = shellSettings;
            S = s;
            _http = httpContextAccessor;
        }


        public override string Name => nameof(DiskWriterTask);

        public override LocalizedString DisplayText => S["Disk Writer Task"];

        public override LocalizedString Category => S["UI"];

        public override IEnumerable<Outcome> GetPossibleOutcomes(WorkflowExecutionContext workflowContext, ActivityContext activityContext)
        {
            return Outcomes(S["Done"]);
        }
        
        public override bool CanExecute(WorkflowExecutionContext workflowContext, ActivityContext activityContext)
        {
            return _http.HttpContext?.Request?.Form != null;
        }

        public override async Task<ActivityExecutionResult> ExecuteAsync(WorkflowExecutionContext workflowContext, ActivityContext activityContext)
        {
            var shell = _shellOptions.Value;
            var directory = PathExtensions.Combine(shell.ShellsApplicationDataPath, shell.ShellsContainerName, _shellSettings.Name, Folder);

            if (!Directory.Exists(directory))
                Directory.CreateDirectory(directory);

            foreach(var file in _http.HttpContext.Request.Form.Files)
            {
                var toSave = PathExtensions.Combine(directory, file.FileName);
                using (var stream = System.IO.File.Create(toSave))
                {
                    await file.CopyToAsync(stream);
                }
            }

            return Outcomes("Done");
        }

        public string Folder
        {
            get => GetProperty<string>();
            set => SetProperty(value ?? string.Empty);
        }
    }

    public class DiskWriterViewModel
    {
        public string Folder { get; set; }
    }

    public class DiskWriterTaskDisplay: ActivityDisplayDriver<DiskWriterTask, DiskWriterViewModel>
    {
        protected override void EditActivity(DiskWriterTask activity, DiskWriterViewModel model)
        {
            model.Folder = activity.Folder;
        }

        protected override void UpdateActivity(DiskWriterViewModel model, DiskWriterTask activity)
        {
            activity.Folder = model.Folder;
        }
    }
```

**Comment by dodyg** [comment]
You will have to fill the rest of the views. 

![layout](https://user-images.githubusercontent.com/1761249/80301226-28abfb00-87a3-11ea-8e58-64d1c3c632fe.png)

**Comment by dodyg** [comment]
![test](https://user-images.githubusercontent.com/1761249/80301251-4aa57d80-87a3-11ea-947d-87e11cd0b601.png)

**Comment by dodyg** [comment]
![form-1](https://user-images.githubusercontent.com/1761249/80301278-804a6680-87a3-11ea-9502-6df83017eea9.png)
![form-2](https://user-images.githubusercontent.com/1761249/80301279-80e2fd00-87a3-11ea-9654-df52a4619aa4.png)

**Comment by dodyg** [comment]
Because the Form widget doesn't support multiple, you have to modify the generated html with Javascript

```
$(function(){
	$('form').attr("enctype","multipart/form-data");
});
```

**Comment by dodyg** [comment]
![disk-writer-edit](https://user-images.githubusercontent.com/1761249/80301350-f5b63700-87a3-11ea-9412-c4c9dbfa845c.png)

**Comment by dodyg** [comment]
This is my first time writing a custom activity task so this one is gonna be rough at the edges.

**Comment by dodyg** [comment]
Simple worklflow that will add data from the form

![property-task](https://user-images.githubusercontent.com/1761249/80302872-f901f080-87ac-11ea-91c8-f4d06fa1dc0e.png)
![property-task-2](https://user-images.githubusercontent.com/1761249/80302874-fa331d80-87ac-11ea-9038-3e4dcabb6d04.png)
![property-task-3](https://user-images.githubusercontent.com/1761249/80302876-fa331d80-87ac-11ea-8f02-4a682176bb54.png)

**Comment by amira-elbatal** [comment]
@dodyg 
can upload your solution ?

**Comment by dodyg** [comment]
There you go. 
[Disk Writer.zip](https://github.com/OrchardCMS/OrchardCore/files/4539596/Disk.Writer.zip)

**Comment by amira-elbatal** [comment]
@dodyg 
thanks.

**Comment by hishamco** [comment]
Thanks a lot @dodyg .. very detailed and useful explanation as usual 🥇

**Comment by duncanhoggan** [comment]
@dodyg I updated your Task to include liquid support on the folder.

[DiskWriterTask.cs](https://gist.github.com/duncanhoggan/173a0aaf1e4c89ba18f84b00db192e75.js)

**Comment by Flarescape** [comment]
I've done something similar around two months ago and maybe I'm allowed to add some thoughts about security and validation.
In my case, I've added some more fields to the file upload activity and some more output options:

**Fields:**

- Allowed extensions: Validate file extensions = Output: Invalid file extension
- Allowed file size: Validate file size = Output: Invalid file size
- Save file to: Saves the file to a specific location e.g. some media folder.

**Optional fields:**

- Invalid file extension message: Integrated Notify and/or Form validation error
- Invalid file size message: Integrated Notify and/or Form validation error
- Default failed message: Integrated Notify and/or Form validation error
- Default success message: Integrated Notify and/or Form validation error

**Feature field:**

- File name fields: A comma separed list of other form fields which values are added to the file name to prevent dublicates or for better identification.

**Comment by stevetayloruk** [comment]
I think the save part needs to change so if someone is using cloud blob storage then it would also work.  Seems like its hardcoded to the local filesystem.

**Comment by dodyg** [comment]
@stevetayloruk Yup. This is just a quick and dirty implementation for my local project purpose.

**Comment by sebastienros** [comment]
What about having such a task by default in the project?
I assume we would need to use a custom abstraction of the FS, such that we could write to a custom folder or store (blob). These files should not be servable by default. Or have an option to store them in a specific media folder, if we want them to be servable.

Maybe it should only use the media service. And we'd need to have a way to use a custom folder that can't serve files publicly. But we'd be able to browse them from the admin.

**Comment by Flarescape** [comment]
@sebastienros This would be really great.
I mean, I'm currently working with my own solution as described, but it's far from perfect and not for everyones needs.
In addition to that, there is a very common scenario, where people are attaching files to a form: a job application.
With that, you have two Options to handle the files:
1.) Store them inside the media/file system
2.) Send them by mail with a workflow

In case of option 2, I've extended the OrchardCore.Email module to support attachments which would be really helpfull as default functionality.
My file upload task sets the paths as workflow variables which the extended email module reads and attaches them to the mail.
This could be solved way better, but it works for now.

The ability to choose wheter a media folder is public or private would be awesome, because I would really like to store files inside the media, that are only available through the admin panel, so I don't have to save my files outside of the media folder with a hardcoded path anymore.

**Comment by stevetayloruk** [comment]
@sebastienros Is there any plans to have media folders/files securable through the permission system?

What would your approach be to achieve this?

**Comment by sebastienros** [comment]
Per role, per folder:
- Write in folder
- List files in folder

In the attached media field, we already do that, by forcing a custom folder, and not showing other ones.

Bonus feature (harder):
- Serve files in folder, by permission. Only serve the file if the current user's role (even anonymous) is allowed to be served the file.

I think it should be done at the folder level to simplify the UI.

**Comment by Piedone** [comment]
The Form widget now supports `multipart/form-data`. If https://github.com/OrchardCMS/OrchardCore/pull/12218 is merged, we'll have file uploads in Workflows too.

The most recent conversation here changed over to authorized Media (file storage) access though. For that, we have this other issue: https://github.com/OrchardCMS/OrchardCore/issues/3590.
